File: eco-api/src/main/java/com/willfp/eco/core/data/keys/KeyRegistry.java
package com.willfp.eco.core.data.keys;

import com.willfp.eco.core.Eco;
import org.bukkit.NamespacedKey;
import org.jetbrains.annotations.ApiStatus;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Set;

/**
 * API to register persistent data keys.
 */
@ApiStatus.Internal
@Eco.HandlerComponent
public interface KeyRegistry {
    /**
     * Register a persistent data key to be stored.
     *
     * @param key The key.
     */
    void registerKey(@NotNull PersistentDataKey<?> key);

    /**
     * Get all registered keys.
     *
     * @return The keys.
     */
    Set<PersistentDataKey<?>> getRegisteredKeys();

    /**
     * Get persistent data key from namespaced key.
     *
     * @param namespacedKey The key.
     * @return The key, or null if not found.
     */
    @Nullable
    PersistentDataKey<?> getKeyFrom(@NotNull NamespacedKey namespacedKey);
}

File: eco-core/core-plugin/src/main/kotlin/com/willfp/eco/internal/spigot/data/KeyRegistry.kt
package com.willfp.eco.internal.spigot.data

import com.willfp.eco.core.data.keys.PersistentDataKey
import org.bukkit.NamespacedKey

object KeyRegistry {
    private val registry = mutableMapOf<NamespacedKey, PersistentDataKey<*>>()

    fun registerKey(key: PersistentDataKey<*>) {
        if (key.defaultValue == null) {
            throw IllegalArgumentException("Default value cannot be null!")
        }

        this.registry[key.key] = key
    }

    fun getRegisteredKeys(): Set<PersistentDataKey<*>> {
        return registry.values.toSet()
    }
}
